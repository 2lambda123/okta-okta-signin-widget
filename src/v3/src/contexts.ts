/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxMessage, IdxTransaction, OktaAuth } from '@okta/okta-auth-js';
import { createContext } from 'preact';
import { MutableRef, StateUpdater, useContext } from 'preact/hooks';

import {
  ActionOptions,
  FormBag,
  WidgetProps,
} from './types';

// Widget context
type IWidgetContext = {
  authClient: OktaAuth;
  widgetProps: WidgetProps;
  setMessages: StateUpdater<IdxMessage[]>;
  // // TODO: OKTA-502849 - Update param type
  // (RenderSuccessCallback / RenderErrorCallback) once merged into okta-signin-widget
  onSuccessCallback?: (data: Record<string, unknown>) => void;
  onErrorCallback?: (data: Record<string, unknown>) => void;
  idxTransaction: IdxTransaction | undefined;
  setIdxTransaction: StateUpdater<IdxTransaction | undefined>;
  stepToRender: string | undefined;
  setStepToRender: StateUpdater<string | undefined>;
  data: Record<string, unknown>;
  setData: StateUpdater<Record<string, unknown>>;
  dataSchemaRef: MutableRef<FormBag['dataSchema'] | undefined>;
};

const createWidgetContext = <T extends unknown>() => {
  // Create a context with a generic parameter or undefined
  const genericContext = createContext<T | undefined>(undefined);

  // Check if the value provided to the context is defined or throw an error
  const useWidgetContext = () => {
    const contextIsDefined = useContext(genericContext);
    if (!contextIsDefined) {
      throw new Error('useWidgetContext must be used within a Provider');
    }
    return contextIsDefined;
  };

  return [useWidgetContext, genericContext.Provider] as const;
};

const [useWidgetContext, WidgetContextProvider] = createWidgetContext<IWidgetContext>();

export { useWidgetContext, WidgetContextProvider };

// Form context
type IFormContext = {
  submissionOptionsRef: MutableRef<ActionOptions | undefined>;
};

export const FormContext = createContext<IFormContext>({} as IFormContext);

export const useFormContext = () => useContext(FormContext);

// Stepper context
type IStepperContext = {
  setStepIndex: StateUpdater<number | undefined>;
};

export const StepperContext = createContext<IStepperContext>({
  setStepIndex: () => {},
});

export const useStepperContext = () => useContext(StepperContext);
