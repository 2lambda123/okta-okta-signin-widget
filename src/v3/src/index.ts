/*
 * Copyright (c) 2022-present, Okta, Inc. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant
 * to the Apache License, Version 2.0 (the "License.")
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { config } from '../widgetrc';
import { OktaSignIn } from './OktaSignIn';

// if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
//   import('./mocks/browser')
//     .then(({ getWorker }) => getWorker())
//     .then((worker) => worker?.start());
// }

if (import.meta.env.MODE === 'development') {
  // if (import.meta.env.VITE_MSW) {
  //   const { worker } = await import('../mocks/browser');
  //   worker.start({ quiet: import.meta.env.VITE_MSW_QUIET });
  // }
  
  // eslint-disable-next-line no-new
  new OktaSignIn(config);
}
