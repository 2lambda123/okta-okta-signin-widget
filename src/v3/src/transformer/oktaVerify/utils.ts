/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxAuthenticator } from '@okta/okta-auth-js';
import set from 'lodash/set';
import { loc } from 'okta';
import {
  DescriptionElement,
  ListElement,
  QRCodeElement,
  StepperNavButtonConfigAttrs,
  StepperNavButtonConfigDirection,
  UISchemaElement,
} from 'src/types';

export const appendDescriptionElements = (
  elements: UISchemaElement[],
  authenticator?: IdxAuthenticator,
): void => {
  if (!authenticator?.contextualData) {
    return;
  }

  const { contextualData } = authenticator;

  if (contextualData.qrcode) {
    const { href } = contextualData.qrcode;
    const { displayName } = authenticator;
    elements.push({
      type: 'List',
      options: {
        items: [
          loc('oie.enroll.okta_verify.qrcode.step1', 'login'),
          loc('oie.enroll.okta_verify.qrcode.step2', 'login'),
          loc('oie.enroll.okta_verify.qrcode.step3', 'login'),
        ],
        type: 'ordered',
      },
    } as ListElement);

    elements.push({
      type: 'QRCode',
      options: {
        label: displayName,
        data: href,
      },
    } as QRCodeElement);
    return;
  }

  // @ts-ignore OKTA-496373 - missing props from interface
  const { email } = contextualData;
  if (email) {
    elements.push({
      type: 'Description',
      options: {
        // TODO: revist this to use oie i18n string (EnrollChannelPollDescriptionView.js)
        content: loc('next.enroll.okta_verify.email.info', 'login', [email]),
      },
    } as DescriptionElement);
    return;
  }

  // @ts-ignore OKTA-496373 - missing props from interface
  const { phoneNumber } = contextualData;
  if (phoneNumber) {
    elements.push({
      type: 'Description',
      options: {
        // TODO: revist this to use oie i18n string (EnrollChannelPollDescriptionView.js)
        content: loc('next.enroll.okta_verify.sms.info', 'login', [phoneNumber]),
      },
    } as DescriptionElement);
  }
};

export const createNavButtonConfig = (hasQrCode: boolean): {
  navButtonsConfig: Record<StepperNavButtonConfigDirection, StepperNavButtonConfigAttrs>,
} => {
  const stepperNavBtnConfig = {
    navButtonsConfig: {
      next: {
        variant: 'secondary',
        label: hasQrCode
          ? 'enroll.totp.cannotScan'
          : 'next.enroll.okta_verify.switch.channel.link.text',
      },
    } as Record<StepperNavButtonConfigDirection, StepperNavButtonConfigAttrs>,
  };

  if (hasQrCode) {
    set(stepperNavBtnConfig, 'navButtonsConfig.prev', {
      variant: 'secondary',
      label: 'next.enroll.okta_verify.switch.channel.link.text',
    } as StepperNavButtonConfigAttrs);
  }

  return stepperNavBtnConfig;
};

export const getTitleKey = (selectedChannel?: string): string => {
  switch (selectedChannel) {
    case 'email':
      return 'oie.enroll.okta_verify.setup.email.title';
    case 'sms':
      return 'oie.enroll.okta_verify.setup.sms.title';
    default:
      return 'oie.enroll.okta_verify.setup.title';
  }
};
