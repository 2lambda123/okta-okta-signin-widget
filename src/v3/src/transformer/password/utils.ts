/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { PasswordWithConfirmationElement, UISchemaElement } from '../../types';

export const getPasswordMatchingKey = (
  data: Record<string, unknown>,
): string | undefined => {
  if ('credentials.passcode' in data) {
    return 'credentials.passcode';
  }

  if ('credentials.password' in data) {
    return 'credentials.password';
  }

  if ('credentials.newPassword' in data) {
    return 'credentials.newPassword';
  }

  // Should never hit this case as it should be one of values defined above
  return undefined;
};

export const updatePasswordWithConfirmationBtnLabel = (
  elements: UISchemaElement[],
  label: string,
): void => {
  const pwEnrollmentEle = elements.find(
    (element) => element.type === 'PasswordWithConfirmation',
  ) as PasswordWithConfirmationElement;
  if (pwEnrollmentEle) {
    pwEnrollmentEle.options = {
      ...pwEnrollmentEle.options,
      ctaLabel: label,
    };
  }
};
