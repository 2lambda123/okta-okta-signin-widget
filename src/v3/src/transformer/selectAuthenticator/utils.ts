/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';
import { IdxOption } from '@okta/okta-auth-js/lib/idx/types/idx-js';
import { loc } from 'okta';

import { AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP, AUTHENTICATOR_KEY } from '../../constants';
import { AuthenticatorButtonElement } from '../../types';

const getAuthenticatorOption = (
  options: IdxOption[],
  authenticatorKey: string,
): IdxOption | undefined => options?.find(
  ({ relatesTo }) => relatesTo?.key === authenticatorKey,
);

export const getOptionValue = (
  inputs: Input[],
  key: string,
): Input | undefined => inputs?.find(
  ({ name }) => name === key,
);

const getAuthenticatorDataSeVal = (authenticatorKey: string, methodType?: string): string => {
  if (authenticatorKey) {
    const method = methodType ? `-${methodType}` : '';
    return `${authenticatorKey}${method}`;
  }
  return '';
};

const buildOktaVerifyOptions = (
  options: IdxOption[],
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => {
  const ovRemediation = getAuthenticatorOption(options, AUTHENTICATOR_KEY.OV);
  const id = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'id')?.value;
  const methodType = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'methodType');
  if (!methodType || !methodType.options?.length) {
    return [];
  }

  return methodType.options.map((option) => ({
    type: 'AuthenticatorButton',
    label: option.label,
    options: {
      key: AUTHENTICATOR_KEY.OV,
      ctaLabel: isEnroll
        ? loc('oie.enroll.authenticator.button.text', 'login')
        : loc('oie.verify.authenticator.button.text', 'login'),
      description: isEnroll && loc(AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[AUTHENTICATOR_KEY.OV], 'login'),
      actionParams: {
        'authenticator.methodType': option.value,
        'authenticator.id': id,
      },
      dataSe: getAuthenticatorDataSeVal(AUTHENTICATOR_KEY.OV, option.value as string),
    },
  })) as AuthenticatorButtonElement[];
};

const getAuthenticatorDescription = (
  options: IdxOption[],
  authenticatorKey: string,
  isEnroll?: boolean,
): string | undefined => {
  if (!authenticatorKey) {
    return undefined;
  }
  if (isEnroll) {
    if (authenticatorKey === AUTHENTICATOR_KEY.ON_PREM) {
      return 'oie.on_prem.authenticator.description';
    }
    return AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[authenticatorKey];
  }

  if (authenticatorKey === AUTHENTICATOR_KEY.PHONE) {
    return getAuthenticatorOption(
      options,
      AUTHENTICATOR_KEY.PHONE,
    )?.relatesTo?.profile?.phoneNumber as string;
  }
  return undefined;
};

const getOnPremDescriptionParams = (
  options: IdxOption[],
  authenticatorKey: string,
  isEnroll?: boolean,
): string[] | undefined => {
  if (!isEnroll || authenticatorKey !== AUTHENTICATOR_KEY.ON_PREM) {
    return undefined;
  }
  const vendorName = getAuthenticatorOption(
    options,
    AUTHENTICATOR_KEY.ON_PREM,
  )?.relatesTo?.displayName || 'oie.on_prem.authenticator.default.vendorName';
  return [vendorName];
};

const formatAuthenticatorOptions = (
  options: IdxOption[],
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => (
  options.map((option: IdxOption) => {
    const authenticatorKey = option.relatesTo?.key as string;
    const id = getOptionValue(option.value as Input[], 'id')?.value;
    const methodType = getOptionValue(option.value as Input[], 'methodType')?.value;
    const enrollmentId = getOptionValue(option.value as Input[], 'enrollmentId')?.value;

    const descr = getAuthenticatorDescription(
      options,
      authenticatorKey,
      isEnroll,
    );
    const descrParams = getOnPremDescriptionParams(
      options,
      authenticatorKey,
      isEnroll,
    );
    return {
      type: 'AuthenticatorButton',
      label: option.label,
      options: {
        key: authenticatorKey,
        ctaLabel: isEnroll
          ? loc('oie.enroll.authenticator.button.text', 'login')
          : loc('oie.verify.authenticator.button.text', 'login'),
        description: loc(descr, descrParams),
        actionParams: {
          'authenticator.id': id,
          'authenticator.methodType': methodType,
          'authenticator.enrollmentId': enrollmentId,
        },
        dataSe: getAuthenticatorDataSeVal(
          authenticatorKey,
          typeof methodType === 'string' ? methodType : undefined,
        ),
      },
    } as AuthenticatorButtonElement;
  })
);

const getAuthenticatorButtonElements = (
  options: IdxOption[],
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => {
  const formattedOptions = formatAuthenticatorOptions(options, isEnroll);

  // appending OV options back to its original spot
  const ovOptions = buildOktaVerifyOptions(options, isEnroll);
  if (ovOptions.length && options?.length) {
    const ovIndex = options.findIndex(({ relatesTo }) => relatesTo?.key === AUTHENTICATOR_KEY.OV);
    formattedOptions.splice(ovIndex, 1, ...ovOptions);
  }

  return formattedOptions;
};

export const getOVMethodTypeAuthenticatorButtonElements = (
  options: IdxOption[],
): AuthenticatorButtonElement[] => {
  if (!options.length) {
    return [];
  }

  return options.map((option) => ({
    type: 'AuthenticatorButton',
    label: option.label,
    options: {
      key: AUTHENTICATOR_KEY.OV,
      ctaLabel: loc('oie.verify.authenticator.button.text', 'login'),
      actionParams: {
        'authenticator.methodType': (option.value as string),
      },
    },
  })) as AuthenticatorButtonElement[];
};

export const isOnlyPushWithAutoChallenge = (
  inputs?: Input[],
): boolean => {
  const methodType = inputs?.find(({ name }) => name === 'methodType');
  const autoChallenge = inputs?.find(({ name }) => name === 'autoChallenge');

  return typeof autoChallenge !== 'undefined'
    && methodType?.options?.length === 1
    && methodType.options[0].value === 'push';
};

export const getAuthenticatorVerifyButtonElements = (
  authenticatorOptions: IdxOption[],
):AuthenticatorButtonElement[] => getAuthenticatorButtonElements(
  authenticatorOptions,
);

export const getAuthenticatorEnrollButtonElements = (
  authenticatorOptions: IdxOption[],
): AuthenticatorButtonElement[] => getAuthenticatorButtonElements(
  authenticatorOptions,
  true,
);
