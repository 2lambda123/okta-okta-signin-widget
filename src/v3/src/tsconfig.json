{
  "extends": "../tsconfig.base.json",
  "include": [
    "./**/PhoneAuthenticator.test.tsx"
  ]
}
