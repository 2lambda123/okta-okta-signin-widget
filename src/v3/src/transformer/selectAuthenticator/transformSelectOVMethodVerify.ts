/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';

import {
  ButtonElement,
  ButtonType,
  DescriptionElement,
  FieldElement,
  IdxStepTransformer,
  TitleElement,
  UISchemaElement,
} from '../../types';
import { loc } from '../../util';
import { getUIElementWithName, removeUIElementWithName } from '../utils';
import { getOVMethodTypeAuthenticatorButtonElements, isOnlyPushWithAutoChallenge } from './utils';

export const transformSelectOVMethodVerify: IdxStepTransformer = ({ transaction, formBag }) => {
  const { nextStep: { inputs } = {} } = transaction;
  const authenticator = inputs?.find(({ name }) => name === 'authenticator') as Input;
  if (!authenticator) {
    return formBag;
  }

  const { uischema, data } = formBag;

  if (isOnlyPushWithAutoChallenge(authenticator.value as Input[])) {
    uischema.elements = removeUIElementWithName(
      'authenticator.methodType',
      uischema.elements as UISchemaElement[],
    );
    const autoChallenge = getUIElementWithName(
      'authenticator.autoChallenge',
      uischema.elements as UISchemaElement[],
    ) as FieldElement;
    if (autoChallenge) {
      data[autoChallenge.options.inputMeta.name] = autoChallenge.options.inputMeta.value;
    }
    data['authenticator.methodType'] = 'push';

    uischema.elements.unshift({
      type: 'Title',
      options: { content: loc('oie.okta_verify.push.title', 'login') },
    } as TitleElement);
    uischema.elements.push({
      type: 'Button',
      label: loc('oie.okta_verify.sendPushButton', 'login'),
      scope: `#/properties/${ButtonType.SUBMIT}`,
      options: {
        type: ButtonType.SUBMIT,
      },
    } as ButtonElement);
  } else {
    const methodType = (authenticator.value as Input[])?.find(({ name }) => name === 'methodType');
    if (!methodType?.options) {
      return formBag;
    }

    const buttonElements = getOVMethodTypeAuthenticatorButtonElements(methodType.options);
    uischema.elements = removeUIElementWithName(
      'authenticator.methodType',
      uischema.elements as UISchemaElement[],
    );
    uischema.elements = uischema.elements.concat(buttonElements);

    const titleElement: TitleElement = {
      type: 'Title',
      options: {
        content: loc('oie.select.authenticators.verify.title', 'login'),
      },
    };
    const descriptionElement: DescriptionElement = {
      type: 'Description',
      options: {
        content: loc('oie.select.authenticators.verify.subtitle', 'login'),
      },
    };

    // Title -> Descr -> Element(s)
    uischema.elements.unshift(descriptionElement);
    uischema.elements.unshift(titleElement);
  }

  return formBag;
};
