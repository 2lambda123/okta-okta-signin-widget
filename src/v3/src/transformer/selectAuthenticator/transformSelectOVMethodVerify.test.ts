/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxOption } from '@okta/okta-auth-js/lib/idx/types/idx-js';
import { IDX_STEP } from 'src/constants';
import { getStubTransactionWithNextStep } from 'src/mocks/utils/utils';
import {
  AuthenticatorButtonElement,
  ButtonElement,
  ButtonType,
  DescriptionElement,
  FieldElement,
  FormBag,
  TitleElement,
  UISchemaLayoutType,
  WidgetProps,
} from 'src/types';

import { transformSelectOVMethodVerify } from './transformSelectOVMethodVerify';

const getMockMethodTypes = (): AuthenticatorButtonElement[] => {
  const authenticators: AuthenticatorButtonElement[] = [];
  authenticators.push({
    type: 'AuthenticatorButton',
    label: 'Get a push notification',
    options: {
      key: 'okta_verify',
      ctaLabel: 'Select',
      actionParams: {
        'authenticator.methodType': 'push',
      },
    },
  });
  authenticators.push({
    type: 'AuthenticatorButton',
    label: 'Enter a code',
    options: {
      key: 'okta_verify',
      ctaLabel: 'Select',
      actionParams: {
        'authenticator.methodType': 'totp',
      },
    },
  });
  return authenticators;
};

let isPushOnly = false;
jest.mock('./utils', () => ({
  getOVMethodTypeAuthenticatorButtonElements: (
    options?: IdxOption[],
  ) => (options?.length ? getMockMethodTypes() : []),
  isOnlyPushWithAutoChallenge: jest.fn().mockImplementation(
    () => isPushOnly,
  ),
}));

describe('Transform Select OV Method Verify Tests', () => {
  const transaction = getStubTransactionWithNextStep();
  let formBag: FormBag;
  const mockProps: WidgetProps = {};

  beforeEach(() => {
    formBag = {
      dataSchema: {},
      schema: {},
      uischema: {
        type: UISchemaLayoutType.VERTICAL,
        elements: [
          {
            type: 'Control',
            name: 'authenticator.methodType',
          } as FieldElement,
          {
            type: 'Control',
            name: 'authenticator.autoChallenge',
            options: { inputMeta: { name: 'authenticator.autoChallenge', value: 'true' } },
          } as FieldElement,
        ],
      },
      data: {},
    };
    transaction.nextStep = {
      name: IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
      inputs: [
        {
          name: 'authenticator',
          value: [{ name: 'autoChallenge', type: 'boolean', value: 'true' }],
        },
      ],
    };
  });

  it('should not transform elements when transaction is missing inputs', () => {
    transaction.nextStep.inputs = undefined;
    expect(transformSelectOVMethodVerify(transaction, formBag, mockProps)).toEqual(formBag);
  });

  it('should not transform elements when transaction is missing methodType from input values', () => {
    transaction.nextStep.inputs = [{ name: 'authenticator' }];
    expect(transformSelectOVMethodVerify(transaction, formBag, mockProps)).toEqual(formBag);
  });

  it('should transform elements when transaction only contains push method type '
    + 'with autoChallenge option', () => {
    isPushOnly = true;
    const updatedFormBag = transformSelectOVMethodVerify(transaction, formBag, mockProps);

    expect(updatedFormBag.uischema.elements.length).toBe(3);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.okta_verify.push.title');
    expect((updatedFormBag.uischema.elements[1] as FieldElement).name)
      .toBe('authenticator.autoChallenge');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).label)
      .toBe('oie.okta_verify.sendPushButton');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
    expect(updatedFormBag.data)
      .toEqual({ 'authenticator.autoChallenge': 'true', 'authenticator.methodType': 'push' });
  });

  it('should transform elements when transaction contains push and totp method types', () => {
    formBag.uischema.elements = [{
      type: 'Control',
      name: 'authenticator.methodType',
      options: {
        inputMeta: {
          name: 'authenticator.methodType',
          options: [{ label: 'Enter code', value: 'totp' }, { label: 'Push', value: 'push' }],
        },
      },
    } as FieldElement];
    transaction.nextStep = {
      name: IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
      inputs: [
        {
          name: 'authenticator',
          value: [{
            name: 'methodType',
            options: [{ label: 'Enter code', value: 'totp' }, { label: 'Push', value: 'push' }],
          }],
        },
      ],
    };
    isPushOnly = false;
    const updatedFormBag = transformSelectOVMethodVerify(transaction, formBag, mockProps);

    expect(updatedFormBag.uischema.elements.length).toBe(4);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.select.authenticators.verify.title');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).type).toBe('Description');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).options?.content)
      .toBe('oie.select.authenticators.verify.subtitle');
    expect((updatedFormBag.uischema.elements[2] as AuthenticatorButtonElement).type).toBe('AuthenticatorButton');
    expect((updatedFormBag.uischema.elements[2] as AuthenticatorButtonElement).label)
      .toBe('Get a push notification');
    expect((updatedFormBag.uischema.elements[3] as AuthenticatorButtonElement).label).toBe('Enter a code');
  });
});
