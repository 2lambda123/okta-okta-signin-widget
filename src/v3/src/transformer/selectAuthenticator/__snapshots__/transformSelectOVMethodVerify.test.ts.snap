// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform Select OV Method Verify Tests should transform elements when transaction contains push and totp method types 1`] = `
Object {
  "data": Object {
    "authenticator.autoChallenge": "true",
    "authenticator.methodType": "push",
  },
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.select.authenticators.verify.title",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "oie.select.authenticators.verify.subtitle",
        },
        "type": "Description",
      },
      Object {
        "options": Object {
          "buttons": Array [
            Object {
              "label": "Get a push notification",
              "options": Object {
                "actionParams": Object {
                  "authenticator.methodType": "push",
                },
                "ctaLabel": "Select",
                "key": "okta_verify",
                "step": "select-authenticator-authenticate",
                "type": "button",
              },
              "type": "AuthenticatorButton",
            },
            Object {
              "label": "Enter a code",
              "options": Object {
                "actionParams": Object {
                  "authenticator.methodType": "totp",
                },
                "ctaLabel": "Select",
                "key": "okta_verify",
                "step": "select-authenticator-authenticate",
                "type": "button",
              },
              "type": "AuthenticatorButton",
            },
          ],
        },
        "type": "AuthenticatorButtonList",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Transform Select OV Method Verify Tests should transform elements when transaction only contains push method type with autoChallenge option 1`] = `
Object {
  "data": Object {
    "authenticator.autoChallenge": "true",
    "authenticator.methodType": "push",
  },
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.okta_verify.push.title",
        },
        "type": "Title",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "authenticator.autoChallenge",
            "value": "true",
          },
        },
        "type": "Field",
      },
      Object {
        "label": "oie.okta_verify.sendPushButton",
        "options": Object {
          "step": "select-authenticator-authenticate",
          "type": "submit",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
